src/cargo/util/process_builder.rs

```diff
@@ -329,6 +347,53 @@ impl ProcessBuilder {
         }
         command
     }
+
+    /// Converts `ProcessBuilder` into a `std::process::Command` and a rustc style response
+    /// file.  Also handles the jobserver, if present.
+    pub fn build_command_and_response_file(&self) -> CargoResult<(Command, tempfile::TempPath)> {
+        // The rust linker also jumps through similar hoops, although with a different
+        // of response file, which this borrows from.  Some references:
+        // https://github.com/rust-lang/rust/blob/ef92009c1dbe2750f1d24a6619b827721fb49749/src/librustc_codegen_ssa/back/link.rs#L935
+        // https://github.com/rust-lang/rust/blob/ef92009c1dbe2750f1d24a6619b827721fb49749/src/librustc_codegen_ssa/back/command.rs#L135
+
+        let mut command = Command::new(&self.program);
+        if let Some(cwd) = self.get_cwd() {
+            command.current_dir(cwd);
+        }
+        // cmd.exe can handle up to 8k work of args, this leaves some headroom if using a .cmd rustc wrapper.
+        let mut cmd_remaining: usize = 1024 * 6;
+        let mut response_file = tempfile::NamedTempFile::new()?;
+        for arg in &self.args {
+            cmd_remaining = cmd_remaining.saturating_sub(arg.len());
+            if cmd_remaining > 0 {
+                command.arg(arg);
+            } else if let Some(arg) = arg.to_str() {
+                writeln!(response_file, "{}", arg)?;
+            } else {
+                return Err(internal(format!(
+                    "argument {:?} contains invalid unicode",
+                    arg
+                )));
+            }
+        }
+        let mut arg = OsString::from("@");
+        arg.push(response_file.path());
+        command.arg(arg);
+        for (k, v) in &self.env {
```

Comment:
This is likely duplicated (as well as the jobserver bit below) with a different method in this struct, can those be deduplicated?